src/helpers/proposed_network.rs

```diff
@@ -0,0 +1,198 @@
+        self.callback
+            .send(query_id)
+            .map_err(|_| NetworkEventError::CallbackFailed {
+                event_name: "CreateQuery",
+                query_id,
+            })
+    }
+}
+
+pub struct PrepareQueryData {
+    pub query_id: QueryId,
+    callback: oneshot::Sender<()>,
+}
+
+impl PrepareQueryData {
+    #[must_use]
+    pub fn new(query_id: QueryId, callback: oneshot::Sender<()>) -> Self {
+        PrepareQueryData { query_id, callback }
+    }
+
+    pub fn respond(self) -> Result<(), NetworkEventError> {
+        let query_id = self.query_id;
+        self.callback
+            .send(())
+            .map_err(|_| NetworkEventError::CallbackFailed {
+                event_name: "PrepareQuery",
+                query_id,
+            })
+    }
+}
+
+pub struct StartMulData {
+    pub query_id: QueryId,
+    pub endpoints_positions: [Uri; 3],
+    pub field_type: String,
+    pub data_stream: ByteArrStream,
+    callback: oneshot::Sender<()>,
+}
+
+impl StartMulData {
+    pub fn new(
+        query_id: QueryId,
+        endpoints_positions: [Uri; 3],
+        field_type: String,
+        data_stream: ByteArrStream,
+        callback: oneshot::Sender<()>,
+    ) -> Self {
+        StartMulData {
+            query_id,
+            endpoints_positions,
+            field_type,
+            data_stream,
+            callback,
+        }
+    }
+
+    pub fn respond(self) -> Result<(), NetworkEventError> {
+        let query_id = self.query_id;
+        self.callback
+            .send(())
+            .map_err(|_| NetworkEventError::CallbackFailed {
+                event_name: "StartMul",
+                query_id,
+            })
+    }
+}
+
+pub struct MulData {
+    pub query_id: QueryId,
+    pub endpoints_positions: [Uri; 3],
+    pub endpoints_to_roles_and_stream: HashMap<Uri, (Role, ByteArrStream)>,
+}
+
+impl MulData {
+    pub fn new(
+        query_id: QueryId,
+        endpoints_positions: [Uri; 3],
+        endpoints_to_roles_and_stream: HashMap<Uri, (Role, ByteArrStream)>,
+    ) -> Self {
+        Self {
+            query_id,
+            endpoints_positions,
+            endpoints_to_roles_and_stream,
+        }
+    }
+}
+
+pub struct RingEventData {
+    pub query_id: QueryId,
+    pub roles_to_endpoints: HashMap<Role, Uri>,
+    pub ring_event: RingEvent,
+}
+
+impl RingEventData {
+    pub fn new(
+        query_id: QueryId,
+        roles_to_endpoints: HashMap<Role, Uri>,
+        ring_event: RingEvent,
+    ) -> Self {
+        Self {
+            query_id,
+            roles_to_endpoints,
+            ring_event,
+        }
+    }
+}
+
+pub enum NetworkCommand {
+    // Commands sent to handle query creation and initialization
+    CreateQuery(CreateQueryData),
+    PrepareQuery(PrepareQueryData),
+    StartMul(StartMulData),
+    Mul(MulData),
+
+    // Commands sent within the context of a `Ring`, to be used internally
+    RingEvent(RingEventData),
+}
+
+pub trait Network<R: Ring> {
```

Comment:
we may need a trait here - it would be great to guarantee for each command to have a `respond` method

Reply:
that's wrt to `NetworkCommand`. alternatively, `NetworkCommand` can define `respond` method

Reply:
i tried to implement that, but ran into some weirdness that seemed overly cumbersome. How do I define the trait such that it can work for any mixture of stored data/output data. by which I mean, each command will store different data, and each will expect different data to respond with. e.g. `CreateQuery` responds with `QueryId`, while `PrepareQuery` responds with `()`

Reply:
that looks like a use case for associated types?

Reply:
i'll...take a shot at it, but i tried to use associated types and couldnt make it look right